pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,481 @@
+					if err != nil {
+						klog.Errorf("failed to update query group: %+v", err)
+					}
+				} else {
+					klog.V(2).InfoS("no member in query group: ", c.queryGroupStatus.localMembers)
+				}
+			} else if event.eType == rulechanged {
+				if len(c.queryGroupStatus.localMembers) > 0 {
+					err := c.updateGroup()
+					if err != nil {
+						klog.Errorf("failed to update query group: %+v", err)
+					}
+				} else {
+					klog.V(2).InfoS("no member in query group: ", c.queryGroupStatus.localMembers)
+				}
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (c *multicastController) groupIsStale() bool {
+	return false
+}
+
+func (c *multicastController) groupHasInstalled() bool {
+	return c.groupInstalled
+}
+
+func (c *multicastController) updateGroup() error {
+	groupKey := mcastAllHosts.String()
+	c.queryGroupStatus.mutex.Lock()
+	defer c.queryGroupStatus.mutex.Unlock()
+	memberPorts := make([]uint32, 0)
+	blocked_ports := make(map[uint32]bool)
+	for memberInterfaceName := range c.queryGroupStatus.localMembers {
+		obj, found := c.ifaceStore.GetInterfaceByName(memberInterfaceName)
+		if !found {
+			klog.InfoS("Failed to find interface from cache", "interface", memberInterfaceName)
+			continue
+		}
+		//*v1alpha1.RuleAction, apitypes.UID, *crdv1beta.NetworkPolicyType, string, error
+		action, _, _, name, _ := c.validation(obj, *mcastAllHostsCIDR, crdv1beta.DirectionIn)
+		if name != "" && (*action == v1alpha1.RuleActionDrop) {
+			klog.V(4).Infof("policy will block ofport: %d, pod: %s/%s", obj.OFPort, obj.PodNamespace, obj.PodName)
+			blocked_ports[uint32(obj.OFPort)] = true
+		}
+		memberPorts = append(memberPorts, uint32(obj.OFPort))
+	}
+	if c.groupHasInstalled() {
+		if c.groupIsStale() {
+			// Remove the multicast flow entry if no local Pod is in the group.
+			if err := c.ofClient.UninstallMulticastFlows(c.queryGroupStatus.group); err != nil {
+				klog.ErrorS(err, "Failed to uninstall multicast flows", "group", groupKey)
+				return err
+			}
+			// Remove the multicast flow entry if no local Pod is in the group.
+			if err := c.ofClient.UninstallGroup(c.queryGroupStatus.ofGroupID); err != nil {
+				klog.ErrorS(err, "Failed to uninstall multicast group", "group", groupKey)
+				return err
+			}
+
+			c.groupInstalled = false
+			klog.InfoS("Removed multicast group from cache after all members left", "group", groupKey)
+			return nil
+		}
+		// Reinstall OpenFlow group because the local pod receivers have changed.
+		if err := c.ofClient.InstallIGMPGroup(c.queryGroupStatus.ofGroupID, blocked_ports, true, memberPorts); err != nil {
+			return err
+		}
+		klog.V(2).InfoS("Updated OpenFlow group for local receivers", "group", groupKey, "ofGroup", c.queryGroupStatus.ofGroupID, "localReceivers", memberPorts)
+		return nil
+	}
+	// Install OpenFlow group for a new multicast group which has local Pod receivers joined.
+	if err := c.ofClient.InstallIGMPGroup(c.queryGroupId, blocked_ports, true, memberPorts); err != nil {
```

Comment:
Do you think it is cleaner to move this OF group flow maintanance for IGMP query to mcast_controller, and let multicastNPController to maintain the group buckets?

Reply:
Yes, done